php-checks/src/main/java/org/sonar/php/checks/utils/FunctionArgumentCheck.java

```diff
@@ -0,0 +1,181 @@
+   * As tree argument the verified argument is given.
+   */
+  protected abstract void createIssue(ExpressionTree tree);
+
+  /**
+   * Several {@link ArgumentVerifier} and {@link ArgumentIndicator} can be included in the check.
+   * This means a single function can be checked for multiple arguments.
+   *
+   * The order of indicators and verifiers must be observed.
+   * On the one hand, an indicator can lead to the termination of the check of a certain function.
+   * On the other hand, issues can be created by verifier before an indicator is triggered.
+   */
+  public void checkArgument(FunctionCallTree tree, String expectedFunctionName, ArgumentIndicator... expectedArgument) {
+    String functionName = CheckUtils.getLowerCaseFunctionName(tree);
+    List<ExpressionTree> arguments = tree.arguments();
+
+    if (expectedFunctionName.equals(functionName)) {
+      for (ArgumentIndicator argumentIndicator: expectedArgument) {
+        if (argumentIndicator.position >= arguments.size() || !verifyArgument(arguments, argumentIndicator)) {
+          return;
+        }
+      }
+    }
+  }
+
+  @Override
+  public void visitCompilationUnit(CompilationUnitTree tree) {
+    assignmentExpressionVisitor = new AssignmentExpressionVisitor(context().symbolTable());
+    tree.accept(assignmentExpressionVisitor);
+    super.visitCompilationUnit(tree);
+  }
+
+  private boolean verifyArgument(List<ExpressionTree> arguments, ArgumentIndicator argumentIndicator) {
+    ExpressionTree argument = arguments.get(argumentIndicator.position);
+    ExpressionTree argumentValue = getAssignedValue(argument);
+
+    Optional<String> value = nameOf(argumentValue);
+    if (value.isPresent()) {
+      String quoteLessLowercaseValue = CheckUtils.trimQuotes(value.get()).toLowerCase(Locale.ENGLISH);
+      boolean containValues = argumentIndicator.values.contains(quoteLessLowercaseValue);
+
+      if (argumentIndicator instanceof ArgumentVerifier && ((ArgumentVerifier) argumentIndicator).raiseIssueOnMatch == containValues) {
+        createIssue(argument);
+        return true;
+      }
+
+      return containValues;
+    }
+
+    return false;
+  }
+
+  private static Optional<String> nameOf(Tree tree) {
+    String name;
+    if (tree instanceof LiteralTree) {
+      name = ((LiteralTree) tree).value();
+    } else {
+      name = CheckUtils.nameOf(tree);
+    }
+    return name != null ? Optional.of(name) : Optional.empty();
+  }
+
+  /**
+   * Try to resolve the value of a variable which is passed as argument.
+   */
+  private ExpressionTree getAssignedValue(ExpressionTree value) {
+    if (value.is(Tree.Kind.VARIABLE_IDENTIFIER)) {
+      Symbol valueSymbol = context().symbolTable().getSymbol(value);
+      return assignmentExpressionVisitor
+        .getUniqueAssignedValue(valueSymbol)
+        .orElse(value);
+    }
+    return value;
+  }
+
+  protected static class ArgumentIndicator {
```

Comment:
I don't understand this name. I don't see the "indication"... I wonder if "ArgumentMatcher" would be a better name.

Reply:
Sounds better